File: stustapay/core/schema/till.py
from typing import Optional

from pydantic import BaseModel

from stustapay.core.schema.user import Privilege, User


class NewTillButton(BaseModel):
    name: str
    product_ids: list[int]


class TillButton(NewTillButton):
    node_id: int
    id: int
    price: float


class NewTillLayout(BaseModel):
    name: str
    description: str
    button_ids: Optional[list[int]] = None
    ticket_ids: Optional[list[int]] = None


class TillLayout(NewTillLayout):
    node_id: int
    id: int


class NewTillProfile(BaseModel):
    name: str
    description: Optional[str] = None
    layout_id: int
    allow_top_up: bool
    allow_cash_out: bool
    allow_ticket_sale: bool
    enable_ssp_payment: bool
    enable_cash_payment: bool
    enable_card_payment: bool


class TillProfile(NewTillProfile):
    node_id: int
    id: int


class NewTill(BaseModel):
    name: str
    description: Optional[str] = None
    active_shift: Optional[str] = None
    active_profile_id: int
    terminal_id: int | None = None


class Till(NewTill):
    node_id: int
    id: int
    z_nr: int

    active_cash_register_id: Optional[int] = None
    current_cash_register_name: Optional[str] = None
    current_cash_register_balance: Optional[float] = None

    tse_id: Optional[int] = None
    tse_serial: Optional[str] = None


class NewCashRegister(BaseModel):
    name: str


class CashRegister(NewCashRegister):
    node_id: int
    id: int
    current_cashier_id: Optional[int]
    current_till_id: Optional[int]
    balance: float
    account_id: int


class NewCashRegisterStocking(BaseModel):
    name: str
    euro200: int = 0
    euro100: int = 0
    euro50: int = 0
    euro20: int = 0
    euro10: int = 0
    euro5: int = 0
    euro2: int = 0
    euro1: int = 0
    cent50: int = 0
    cent20: int = 0
    cent10: int = 0
    cent5: int = 0
    cent2: int = 0
    cent1: int = 0
    variable_in_euro: float = 0.0


class CashRegisterStocking(NewCashRegisterStocking):
    node_id: int
    id: int
    total: float


class UserRoleInfo(BaseModel):
    id: int
    name: str
    is_privileged: bool
    privileges: list[Privilege]

    node_id: int
    node_name: str

    is_at_current_node: bool


class UserInfo(User):
    user_tag_uid: int
    cash_register_id: Optional[int] = None
    cash_register_name: Optional[str] = None
    cash_drawer_balance: Optional[float] = None
    transport_account_balance: Optional[float] = None

    assigned_roles: list[UserRoleInfo]

File: web/libs/models/src/lib/till.ts
import { z } from "zod";

export const NewTillButtonSchema = z.object({
  name: z.string(),
  product_ids: z.array(z.number().int()),
});

export type NewTillButton = z.infer<typeof NewTillButtonSchema>;

export const UpdateTillButtonSchema = z.object({
  ...NewTillButtonSchema.shape,
  id: z.number().int(),
});

export type UpdateTillButton = z.infer<typeof UpdateTillButtonSchema>;

export const TillButtonSchema = z.object({
  ...UpdateTillButtonSchema.shape,
  price: z.number(),
});

export type TillButton = z.infer<typeof TillButtonSchema>;

export const NewTillLayoutSchema = z.object({
  name: z.string(),
  description: z.string().nullable().optional(),
  button_ids: z.array(z.number().int()).nullable().optional(),
  ticket_ids: z.array(z.number().int()).nullable().optional(),
});

export type NewTillLayout = z.infer<typeof NewTillLayoutSchema>;

export const TillLayoutSchema = z.object({
  ...NewTillLayoutSchema.shape,
  id: z.number().int(),
});
export type TillLayout = z.infer<typeof TillLayoutSchema>;

export const NewTillProfileSchema = z.object({
  name: z.string(),
  description: z.string().nullable().optional(),
  layout_id: z.number().int(),
  allow_top_up: z.boolean(),
  allow_cash_out: z.boolean(),
  allow_ticket_sale: z.boolean(),
  allow_ticket_vouchers: z.boolean(),
  enable_ssp_payment: z.boolean(),
  enable_cash_payment: z.boolean(),
  enable_card_payment: z.boolean(),
});

export type NewTillProfile = z.infer<typeof NewTillProfileSchema>;

export const TillProfileSchema = z.object({
  ...NewTillProfileSchema.shape,
  id: z.number().int(),
});
export type TillProfile = z.infer<typeof TillProfileSchema>;

export const NewTillSchema = z.object({
  name: z.string(),
  description: z.string().nullable().optional(),
  active_shift: z.string().nullable().optional(),
  active_profile_id: z.number(),
  active_user_id: z.number().nullable().optional(),
  terminal_id: z.number().nullable().optional(),
});

export type NewTill = z.infer<typeof NewTillSchema>;

export const UpdateTillSchema = z.object({ ...NewTillSchema.shape, id: z.number() });

export type UpdateTill = z.infer<typeof UpdateTillSchema>;

export const NewCashRegisterStockingSchema = z.object({
  name: z.string(),
  euro200: z.number().int().optional(),
  euro100: z.number().int().optional(),
  euro50: z.number().int().optional(),
  euro20: z.number().int().optional(),
  euro10: z.number().int().optional(),
  euro5: z.number().int().optional(),
  euro2: z.number().int().optional(),
  euro1: z.number().int().optional(),
  cent50: z.number().int().optional(),
  cent20: z.number().int().optional(),
  cent10: z.number().int().optional(),
  cent5: z.number().int().optional(),
  cent2: z.number().int().optional(),
  cent1: z.number().int().optional(),
  variable_in_euro: z.number().optional(),
});

export type NewCashRegisterStocking = z.infer<typeof NewCashRegisterStockingSchema>;

export const UpdateCashRegisterStockingSchema = z.object({
  ...NewCashRegisterStockingSchema.shape,
  id: z.number().int(),
});

export type UpdateCashRegisterStocking = z.infer<typeof UpdateCashRegisterStockingSchema>;

export const CashRegisterStockingSchema = z.object({
  ...UpdateCashRegisterStockingSchema.shape,
  total: z.number(),
});

export type CashRegisterStocking = z.infer<typeof CashRegisterStockingSchema>;

export const NewCashRegisterSchema = z.object({
  name: z.string(),
});

export type NewCashRegister = z.infer<typeof NewCashRegisterSchema>;

export const UpdateCashRegisterSchema = NewCashRegisterSchema;

export type UpdateCashRegister = z.infer<typeof UpdateCashRegisterSchema>;

export const CashRegisterSchema = z.object({
  ...NewCashRegisterSchema.shape,
  id: z.number().int(),
  current_cashier_id: z.number().int().nullable(),
  current_till_id: z.number().int().nullable(),
  current_balance: z.number(),
});

export type CashRegister = z.infer<typeof CashRegisterSchema>;
